src/include/optimizer/stats/histogram.h

```diff
@@ -0,0 +1,298 @@
+          i += 1;
+        }
+        double u_j;
+				// TODO: Original paper does not cover this case!
+        if (i == 0) {
+          u_j = s;
+        } else {
+          Bin b_i = bins[i - 1];
+          Bin b_i1 = bins[i];
+          double p_i = b_i.p;
+          double p_i1 = b_i1.p;
+          double m_i = b_i.m;
+          double m_i1 = b_i1.m;
+
+          double a, b, c, d, z;
+          d = s - sum_i;
+          a = m_i1 - m_i;
+          b = 2 * m_i;
+          c = -2 * d;
+          if (a == 0) {
+            z = -c / b;
+          } else {
+            z = (-b + std::sqrt(b * b - 4 * a * c)) / (2 * a);
+          }
+          u_j = p_i + (p_i1 - p_i) * z;
+        }
+        res.push_back(u_j);
+      }
+    }
+    // PrintUniform(res);
+    return res;
+  }
+
+  // Utility function to visualize histogram
+  void Print() {
+    LOG_INFO("Histogram: total_count=[%f] num_bins=[%lu]", total_count, bins.size());
+    for (Bin b : bins) {
+      b.Print();
+    }
+  }
+
+  void PrintUniform(const std::vector<double> &vec) {
+    LOG_INFO("Printing uniform histogram...");
+    for (double x : vec) {
+      LOG_INFO("[%f]", x);
+    }
+  }
+
+ private:
+
+  void InsertBin(Bin &bin) {
+    total_count += bin.m;
+    if (bin.p < minimum) {
+      minimum = bin.p;
+    }
+    if (bin.p > maximum) {
+      maximum = bin.p;
+    }
+
+    int index = binary_search(bins, 0, bins.size() - 1, bin);
+
+    if (index >= 0) {
+      bins[index].m += bin.m;
+    } else {
+      index = std::abs(index) - 1;
+      bins.insert(bins.begin() + index, bin);
+    }
+  }
+
+  // Merge n + 1 number of bins to n bins based on update algorithm.
+  void MergeTwoBinsWithMinGap() {
+    int min_gap_idx = -1;
+    double min_gap = DBL_MAX;
+    for (uint8_t i = 0; i < bins.size() - 1; i++) {
+      double gap = std::abs(bins[i].p - bins[i + 1].p);
+      if (gap < min_gap) {
+        min_gap = gap;
+        min_gap_idx = i;
+      }
+    }
+		// assert(min_gap_idx >= 0 && min_gap_idx < bins.size());
+    Bin &prev_bin = bins[min_gap_idx];
+    Bin &next_bin = bins[min_gap_idx + 1];
+    prev_bin.MergeWith(next_bin);
+    bins.erase(bins.begin() + min_gap_idx + 1);
+  }
+
+  // A more useful binary search based on Java's Collections.binarySearch
+  // It returns -index - 1 if not found, where index is position for insertion:
+  // the first index with element **greater** than the key.
+  template <typename T>
+  int binary_search(const std::vector<T> &vec, int start, int end, const T &key) {
+    if (start > end) {
+        return -(start + 1);
+    }
+
+    const int middle = start + ((end - start) / 2);
+
+    if (vec[middle] == key) {
+        return middle;
+    } else if (vec[middle] > key) {
+        return binary_search(vec, start, middle - 1, key);
+    }
+    return binary_search(vec, middle + 1, end, key);
+  }
+
+ public:
+
+ /*
+  * Represent a bin of a historgram.
+  */
+  class Bin {
+   public:
+
+    double p;
+    double m;
```

Comment:
I guess p stands for position and m stands for something like weight. I am not sure if it is a convention. If not, maybe it would be better to have some comments describing p and m.

Does m always take int value? Seems m will often be multiplied by p, so save it as double will not need further casting. Sometimes m will be added to another m, this may be easier for int.  

Maybe p and m can be private. Seems they are only modified by the helper functions like MergeWith().